### PR TITLE
Evolved operator functions

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Part of #13046: The `evolved_operator_ansatz` and `qaoa_ansatz` functions.

This is based on #13295, so that PR will need merging first.

### Details and comments

* After #13295, the `EvolvedOperatorAnsatz` is already pretty performant since it does the Pauli evolution in Rust. However, with these functions we still gain a speedup since we can do all evolution at once in Rust and don't need go back and forth.

    This benchmarks the construction of an evolved operator ansatz for an Ising Hamiltonian on a square grid with side length ``n``
(so ``n ** 2`` qubits), which has 2 operators; the ZZ-interactions and the X-field. For 100 qubits we get a speedup of 1.33x,
which is increasing with the number of qubits (e.g. we get 1.83x at 10'000 qubits).

    <details closed>

    ```
    -- n = 10, speedup 1.33x
    main: 0.033 +- 0.014
    this: 0.025 +- 0.009

    -- n = 50, speedup 1.64x
    main: 6.218 +- 0.200
    this: 3.792 +- 0.105

    -- n = 100, speedup 1.83x
    main: 100.487 +- 2.275
    this: 55.007 +- 1.139
    ```
    </details>

* This also adds a new function called `hamiltonian_variational_ansatz` (HVA), which automatically splits an input Hamiltonian into commuting terms. This is slightly different to `evolved_operator_ansatz`, which requires you to manually do this split, but IMO the HVA is the form most users would be interested in. We could've changed the default behavior of `evolved_operator_ansatz` to do the splitting, but I didn't want to change behavior for a class that users are already used to.

### CODE REVIEW

**Comment by Cryoris** [comment]
I'm not sure why this fails, the docs build is complaining about 
```
/home/vsts/work/1/s/docs/apidoc/circuit_library.rst:: ERROR: Anonymous hyperlink mismatch: 1 references but 0 targets.
```
My best guess so far is that it doesn't like that the `evolved_operator_ansatz` function has the same name as the module its coming from?

File: releasenotes/notes/clib-evolved-ops-e91c00964c0209ce.yaml
Comment (alexanderivrii) [inline releasenotes/notes/clib-evolved-ops-e91c00964c0209ce.yaml]: typo in "performant"

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (alexanderivrii) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py]:
```suggestion
    For a Hamiltonian :math:`H = \sum_{k=1}^K H_k` where the terms :math:`H_k` consist of only
```

**Review by alexanderivrii — COMMENTED** [review]
I think this is in a good shape. Though I would like to take another look after #13295 is merged.

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: Fixed in 6bf40b8d2

File: releasenotes/notes/clib-evolved-ops-e91c00964c0209ce.yaml
Comment (Cryoris) [reply releasenotes/notes/clib-evolved-ops-e91c00964c0209ce.yaml]: Fixed in 6bf40b8d2

File: crates/accelerate/src/circuit_library/utils.rs
Comment (alexanderivrii) [inline crates/accelerate/src/circuit_library/utils.rs]: I like this very much! Do you think it would make sense / be possible to also use these definition from within the `synthesis` directory? Or should we move these to an even more central place?

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (alexanderivrii) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: A few questions about this. In the formula, ``R`` corresponds to ``reps`` from the docstring? The $\theta_{j, r}$ actually depend on $r$? What is the relation between `evolved_operator_ansatz` and the `PauliEvolutionGate`? And in particular, why in `evolved_operator_ansatz` we have $e^{i\theta O_j}$ and in the definition of `PauliEvolutionGate` we have $e^{-it H}$, are the signs consistent? (My questions really show my ignorance of the subject.)

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (alexanderivrii) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py:89]: Why`initial_state` is not an argument anymore?

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (alexanderivrii) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py:124]: Oh no, the iterators have spilled over from the Rust code into the Python code as well 😨.

File: test/python/circuit/library/test_pauli_feature_map.py
Comment (alexanderivrii) [inline test/python/circuit/library/test_pauli_feature_map.py]: this print can be removed

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (alexanderivrii) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py:97]: Previously having no iterators threw a `ValueError`, correct? Though I think the change is fine.

**Review by alexanderivrii — COMMENTED** [review]
This looks good. I have left a few minor in-place comments and questions.

**Comment by alexanderivrii** [comment]
@Cryoris, one question just came up in a slack discussion with @raynelfss: why do you need the Rust changes at all? I do like the (small) cleanup that moves `Instruction` and `StandardInstruction` to some common place, and I don't mind this being a part of this PR, but we technically don't need it, correct? Could you also please comment why we need the change to `radd_param`.

File: crates/accelerate/src/circuit_library/utils.rs
Comment (Cryoris) [reply crates/accelerate/src/circuit_library/utils.rs]: Yeah I think that would make sense, though we might want to discuss the name a bit if we move this to a central location, so we don't mistake it among all the other instructions 😄

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: * yes, $R$ is `reps`: I'll add a comment
* `evolved_operator_ansatz` uses layers of `PauliEvolutionGate` to implement a sequence of parameterized evolutions. Indeed they use the _same_ sign, but this is wrongly documented 🙂 
* (you didn't ask but for completeness)`hamiltonian_variational_ansatz` is a special type of `evolved_operator_ansatz` where the operators are set to the commuting groups of a Hamiltonian

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py:89]: Because one can simply do
```python
initial_state.compose(evolved_operator_ansatz(...))
```
Before this was an argument as the `EvolvedOperatorAnsatz` was a blueprint and could be reconstructed, in which case we needed to re-attach the initial state. Not being a blueprint anymore, this is no longer needed and we can simplify the code.

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py:97]: Ha, nothing sneaks past you 😛  Though e.g. the n-local circuits would just be empty if no blocks are set. So I used this chance to make the functions more consistent 🙂

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py:124]: Iter-ception! Which world do we iterate in? 🫨

**Comment by Cryoris** [comment]
The changes to Rust are artifacts from developing off a non-merged PR 😄 The Rust changes are not needed, I'll revert them.

File: test/python/circuit/library/test_pauli_feature_map.py
Comment (Cryoris) [reply test/python/circuit/library/test_pauli_feature_map.py]: removed in 3e009a5e7

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: Added the above in 3e009a5e7

File: crates/accelerate/src/circuit_library/utils.rs
Comment (Cryoris) [reply crates/accelerate/src/circuit_library/utils.rs]: As per the comment below, all the Rust changes are reverted in 3e009a5e7

File: qiskit/circuit/library/n_local/qaoa_ansatz.py
Comment (ShellyGarion) [inline qiskit/circuit/library/n_local/qaoa_ansatz.py:42]: might be good to an some more explanation or example here

**Review by ShellyGarion — COMMENTED** [review]
LGTM. I have a minor comment about the docs.

File: qiskit/circuit/library/n_local/qaoa_ansatz.py
Comment (ShellyGarion) [reply qiskit/circuit/library/n_local/qaoa_ansatz.py:42]: Another comment - please note that these functions do not appear in the table with the classes, as in #13310

File: crates/accelerate/src/circuit_library/pauli_evolution.rs
Comment (raynelfss) [inline crates/accelerate/src/circuit_library/pauli_evolution.rs]: Maybe this line was supposed to be removed or back on line 21?

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (raynelfss) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: Since we're specifying types here, should we also say that it returns a `QuantumCircuit`?
```suggestion
def evolved_operator_ansatz(
    operators: BaseOperator | Sequence[BaseOperator],
    reps: int = 1,
    evolution: EvolutionSynthesis | None = None,
    insert_barriers: bool = False,
    name: str = "EvolvedOps",
    parameter_prefix: str | Sequence[str] = "t",
    remove_identities: bool = True,
    flatten: bool | None = None,
) -> QuantumCircuit:
```

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (raynelfss) [inline qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: You could add the return type here as well
```suggestion
) -> QuantumCircuit :
```

File: qiskit/circuit/library/n_local/qaoa_ansatz.py
Comment (raynelfss) [inline qiskit/circuit/library/n_local/qaoa_ansatz.py]:
```suggestion
) -> QuantumCircuit:
```

**Review by raynelfss — COMMENTED** [review]
I'm not too familiar with PauliEvolution or what the `evolved_operator_ansatz` tries to achieve specifically but the logic here is sound to me. I'll let @alexanderivrii have the last word on approving, but great job regardless!

File: qiskit/circuit/library/n_local/qaoa_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/qaoa_ansatz.py]: done in 076f467

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: done in 076f467

File: qiskit/circuit/library/n_local/evolved_operator_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/evolved_operator_ansatz.py]: done in 076f467

File: crates/accelerate/src/circuit_library/pauli_evolution.rs
Comment (Cryoris) [reply crates/accelerate/src/circuit_library/pauli_evolution.rs]: done in 076f467

File: qiskit/circuit/library/n_local/qaoa_ansatz.py
Comment (Cryoris) [reply qiskit/circuit/library/n_local/qaoa_ansatz.py:42]: done in 076f467

**Review by alexanderivrii — DISMISSED** [review]
LGTM!

**Review by raynelfss — APPROVED** [review]
Looks okay, we'll fix the removals with #13407

**Pull request merged**
